hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -77,6 +78,13 @@ public Boolean eval(Row row, ExpressionEvalContext context) {
             return null;
         }
 
+        Class<?> leftClass = left.getClass();
+        Class<?> rightClass = right.getClass();
+
+        if (!(left instanceof Comparable && leftClass.equals(rightClass))) {
+            throw QueryException.error("trying to compare two incomparable objects");
```

Comment:
It is not very obvious from the message what does "incomparable" means. Better messages could be (rough examples):
1. If the object is not comparable: `Cannot compare OBJECT value because it doesn't implement Comparable interface`
2. If objects are of different types: `Cannot compare two OBJECT values, because they have different classes`